[e2e] Install specific version of WC using WC Beta Tester

Stojdza opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR aims to introduce the option to[ install a specified version of WC during test execution](https://github.com/woocommerce/woocommerce/issues/49695) using WC Beta Tester API. If enabled, by default, it will install the latest version of WC (latest final, Beta, or RC). 

More info on how WC Beta Tester can be used to install a specific version of WC can be found in this PR https://github.com/woocommerce/woocommerce/pull/54103.

This functionality will be used for running tests on external environments where installing WC using WC Beta Tester is enabled by default. For local execution and `wp-env`, there should be no changes. 

Part of https://github.com/woocommerce/woocommerce/issues/49695

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Besides the new logic for installing WC using the WC Beta Tester, this PR enables installation only for the Pressable website. 
1. Log in to the Pressable website and check the installed version of WooCommerce. If not, install and activate the latest stable version (we need this for testing).
2. Checkout this branch and run `pnpm test:e2e:pressable basic.spec.js` and wait for the tests to finish. Was it successful?
3. Go to `/wp-admin/plugins.php` page on the Pressable website. The latest release, in this case `[9.6.1](https://github.com/woocommerce/woocommerce/releases/tag/9.6.1)`, should be installed and activated
4. Now run `WC_VERSION=9.6.0-rc.1 pnpm test:e2e:pressable basic.spec.js --ui` and wait for the tests to finish. Was it successful?
5. Go to `/wp-admin/plugins.php` page on the Pressable website. WC `9.6.0-rc.1` should be installed and activated. 
6. Go to `/wp-admin/plugins.php` and deactivate WC (make sure that no WC plugin is active)
7.  Now run `WC_VERSION=9.5.2 pnpm test:e2e:pressable basic.spec.js --ui` and wait for the tests to finish. Was it successful?

Additional testing and consideration:
1. Is CI green? These changes should not affect CI execution and execution with `wp-env`
2. Take a look at `site.setup.js` file, does the new fixture `wcbtApi` makes sense?
3. I find running command with `WC_VERSION=9.5.2 pnpm test:e2e:...` really handy. That way, I could easily run tests against previous versions of WC if I need to check or compare something. Is this something we can use for `wp-env` as well? In that case, we would need to run `wp-env` with WC Beta Tester installed. 

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Install a specific version of WC using WC Beta Tester.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Commit:
Set INSTALL_WC env var

Commit:
Introduce wc-install.setup.js

Commit:
playwright.config.js - Install WC (if set) before 'consumer token setup'

Commit:
Merge branch 'e2e/preinstall-wc' of https://github.com/woocommerce/woocommerce into e2e/preinstall-wc

Commit:
Merge branch 'trunk' into e2e/preinstall-wc

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/fixtures/wc-install.setup.js:
```diff
@@ -0,0 +1,98 @@
+setup.use( { storageState: ADMIN_STATE_PATH } );
+
+setup( 'install WC using WC Beta Tester', async ( { page } ) => {
+	if ( process.env.INSTALL_WC === 'true' ) {
```
May I suggest to use `setup.skip()` instead of this conditional? With [the skip option](https://playwright.dev/docs/api/class-test#test-skip) the setup script will be reported as such and it will be more clear that the plugins was not installed.

Something in the lines of:
```
setup.skip( !process.env.INSTALL_WC, 'Skipping installing the plugin, INSTALL_WC not found');
```

Stojdza replied:
Thanks Adi! This is nice and clean 👍 Implemented it here [b720af5](https://github.com/woocommerce/woocommerce/pull/55063/commits/b720af509827e87a5430e0ae656112afbc923c49)

Stojdza replied:
This is now part of [`site.setup.js` fixture](https://github.com/woocommerce/woocommerce/pull/55063/files#diff-f6a4141a21bfc84664a41827dcf301bc2facc6822d2d1d56ec5eaf6ce6bb5188R109-R112).

Commit:
Use `setup.skip` instead of `if` loop

Commit:
Create `wcbtApi` fixture in fixtures.js

Commit:
Add `Install WC using WC Beta Tester` to site.setup.js

Commit:
Delete file wc-install.setup.js

Commit:
playwright.config.js - Revert to the version from `trunk`

Commit:
Merge branch 'trunk' into e2e/preinstall-wc

Stojdza commented:
@adimoldovan, thanks for pushing me in this direction! 🙌 

> Those endpoints should be available to call as a regular api, no browser should be needed. Maybe we should add a new [fixture](https://github.com/woocommerce/woocommerce/blob/1e4361c10153f19c4fa89d1e79971188055bc470/plugins/woocommerce/tests/e2e-pw/fixtures/fixtures.js#L23), wcbtApi, just like we have for wcAdminApi, that uses the wc-admin-test-helper namespace.

Because of [this comment](https://github.com/woocommerce/woocommerce/pull/54103#issuecomment-2569027694) and how we tested the functionality so far, I thought it was mandatory to load the page first (similar to what we're doing in `token.setup.js`). But thanks to your proposal, I saw it can be done using the browser, which is faster and more reliable. 

In the last few commits, I addressed the feedback and updated the PR description. When you have a time, can you take one more look? 

There is a case when this solution won't work on the external environments - if there's no WC installed on the website, `token.setup.js` will fail. With this "only API" approach, we need a token to install the desired WC version; and we need WC installed to set up a token 🔁 I'm working on a solution at the moment, and I can push it here or in the follow-up PR. It's a case that won't happen often, but I would rather have it done as robust as possible.

adimoldovan commented:
Thanks for applying the suggestions @Stojdza.

> There is a case when this solution won't work on the external environments - if there's no WC installed on the website, `token.setup.js` will fail. With this "only API" approach, we need a token to install the desired WC version; and we need WC installed to set up a token 🔁 

We don't need the token to install the desired WC version, `wcbtApi` is using basic authentication and we have the user and password at that point.

But still, `token.setup.js` will fail if it runs without WC installed, the same as all the other steps in the setup that rely on the plugins being installed.

The install needs to be its own setup file, and it should run before everything else. With the current structure, the global authentication `auth.setup.js` which is the first one running should depend on (the new) `install-wc.setup.js`.

Commit:
Put back install-wc.setup.js and execute it first in `setupProjects`

Commit:
Merge branch 'e2e/preinstall-wc' of https://github.com/woocommerce/woocommerce into e2e/preinstall-wc

Commit:
Merge branch 'trunk' into e2e/preinstall-wc

Stojdza commented:
Thanks for the suggestion @adimoldovan 👍 I added [the changes](https://github.com/woocommerce/woocommerce/pull/55063/commits/14cecaecd597b39d5ed5931cda0c409736b37e4d), and while tests and WC installation step are passing locally against Pressable and wp-env, they are failing on GH with [this error](https://github.com/woocommerce/woocommerce/actions/runs/13198620538/job/36845435821?pr=55063). Looking at why is this happening 👀

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/fixtures/fixtures.js:
```diff
@@ -51,6 +51,88 @@ exports.test = base.test.extend( {
+				return await response.json();
+			},
+
+			/**
+			 * Install a specific WC version.
+			 *
+			 * @param {string} resolvedVersion - The WooCommerce version (e.g., '9.6.0-beta').
+			 */
+			installSpecificVersion: async ( resolvedVersion ) => {
+				const downloadUrl = `https://github.com/woocommerce/woocommerce/releases/download/${ resolvedVersion }/woocommerce.zip`;
+				return await wcbtApi.fetch(
+					'./wp-json/wc-admin-test-helper/live-branches/install/v1',
+					{
+						method: 'POST',
+						data: {
+							pr_name: resolvedVersion,
+							download_url: downloadUrl,
+							version: resolvedVersion,
+						},
+					}
+				);
+			},
+
+			/**
+			 * Activate a specific WC version.
+			 *
+			 * @param {string} resolvedVersion - The WooCommerce version (e.g., '9.6.0-beta').
+			 */
+			activateWC: async ( resolvedVersion ) => {
+				return await wcbtApi.fetch(
+					'./wp-json/wc-admin-test-helper/live-branches/activate/v1',
+					{
+						method: 'POST',
+						data: {
+							version: resolvedVersion,
+						},
+					}
+				);
+			},
+		} );
```
You don't need to implement all these methods here (not even sure that works?), all you need is to use the `wcbtApi`, to expose it to wherever the fixture is used.

The code from the functions `deactivateWC`, `installLatest`, etc. should be in the `install-ws.setup.js` only. I don't think we'll need this anywhere else, so I don't think extracting the logic in public functions is required.

```suggestion
		await use( wcbtApi );
```

Stojdza replied:
Done in [de3a60a](https://github.com/woocommerce/woocommerce/pull/55063/commits/de3a60a5e850f95db30cd9b6eacb7a0648c28fea)

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/playwright.config.js:
```diff
@@ -62,10 +62,16 @@ if ( process.env.CI ) {
 
 export const setupProjects = [
+	{
+		name: 'install wc with wc beta tester',
```
Nitpick: I would go with a shorter title, like "install wc". I don't see it relevant to specify it is using the beta tester.

Stojdza replied:
The shorter the better 👍 I went with beta tester part because I was thinking about someone who might run the tests (let's say against wp-env) and see the step "install wc" skipped, and might think why WC is skipped and not installed

Stojdza replied:
Done - [225b1ac](https://github.com/woocommerce/woocommerce/pull/55063/commits/225b1ac3d70428ca4c52f7a3047b6c273f9098bf)

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/fixtures/install-wc.setup.js:
```diff
@@ -0,0 +1,68 @@
+ * Internal dependencies
+ */
+import { test as setup } from './fixtures';
+import { ADMIN_STATE_PATH } from '../playwright.config';
+
+setup.use( { storageState: ADMIN_STATE_PATH } );
```
This is not needed and it's the cause of the failure you mentioned.

```suggestion
```

adimoldovan reviewed (changes requested):
>...they are failing on GH with [this error](https://github.com/woocommerce/woocommerce/actions/runs/13198620538/job/36845435821?pr=55063).

That's caused by the `setup.use( { storageState: ADMIN_STATE_PATH } );` in the `install-wc.setup.js`. That requires the `./state/admin.json` file to existst, but that is only created by the authentication setup step. You need to delete that part, is is not required here. I also left a code suggestion for this.

> and while tests and WC installation step are passing locally

Deleting the `./state/admin.json` before will cause a failure locally as well. You have that file from previous runs, that's why it's not failing.

Commit:
Remove `ADMIN_STATE_PATH` usage

Commit:
Move methods out of `wcbtApi` fixture

Commit:
Rename setup name - `install wc`

Commit:
Merge branch 'trunk' into e2e/preinstall-wc

Commit:
Add dependencies for Pressable reset

Commit:
Merge branch 'e2e/preinstall-wc' of https://github.com/woocommerce/woocommerce into e2e/preinstall-wc

Commit:
Remove ADMIN_STATE_PATH from reset.setup.js

Commit:
Merge branch 'trunk' into e2e/preinstall-wc

Commit:
Merge branch 'e2e/preinstall-wc' of https://github.com/woocommerce/woocommerce into e2e/preinstall-wc

Commit:
Revert "Remove ADMIN_STATE_PATH from reset.setup.js"

This reverts commit fe4f5a1047999fdf51be7e2b941981122a6d1e04.

Commit:
Add dependencies for WPCOM reset

Stojdza commented:
Feedback addressed. Additionally I updated dependencies for [Pressable](https://github.com/woocommerce/woocommerce/pull/55063/commits/070cadd1254848009e0b8bb3960ac026e9b56c10) and [WPCOM](https://github.com/woocommerce/woocommerce/pull/55063/commits/42cf1c71445ae5cca39117e15cce7772b6cac0d1), since site reset was failing after these changes. 

[Here](https://github.com/woocommerce/woocommerce/actions/runs/13204482197) is the on-demand job with the latest changes. For Pressable, `[install wc] › ../fixtures/install-wc.setup.js:6:6 › Install WC using WC Beta Tester (6.0s)` was successfully executed (also verified WC version on the website), and for WPCOM this step was skipped.

rodelgc commented:
Hello @Stojdza I'm not sure if it's related to the changes in this PR, but I'm seeing multiple WooCommerce versions on the Pressable site.

<img width="1056" alt="Screenshot 2025-02-07 at 10 50 25 PM" src="https://github.com/user-attachments/assets/3de60dc6-d0b9-4511-94e3-dfdf33d0de69" />

Stojdza commented:
@rodelgc 👋 Yes, it was me, testing. With these changes, we don't delete plugins. We can activate/deactivate them and install them. Feel free to remove any of these if needed. Thanks!

adimoldovan commented:
> Additionally I updated dependencies for [Pressable](https://github.com/woocommerce/woocommerce/pull/55063/commits/070cadd1254848009e0b8bb3960ac026e9b56c10) and [WPCOM](https://github.com/woocommerce/woocommerce/pull/55063/commits/42cf1c71445ae5cca39117e15cce7772b6cac0d1), since site reset was failing after these changes.

@Stojdza  I'm not seeing why this is needed. Both site-setup and reset should work fine. Can you point me to such a failure?

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/envs/default-pressable/playwright.config.js:
```diff
@@ -2,6 +2,7 @@ let config = require( '../../playwright.config.js' );
 const { tags } = require( '../../fixtures/fixtures' );
 
 process.env.IS_PRESSABLE = 'true';
+process.env.INSTALL_WC = 'true';
```
This is missing in the wpcom config. Is it intentional?

Stojdza replied:
Yes, because WC on our WPCOM plan cannot be deactivated. I shared this in the Slack update here: p1736022424909529-slack-C0833T6CFHP, and @rodelgc confirmed it in the following message. In this PR, I focused on implementing the functionality and the Pressable website.

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/fixtures/install-wc.setup.js:
```diff
@@ -0,0 +1,114 @@
+	// Install WC
+	if ( wcVersion === 'latest' ) {
+		try {
+			const latestResponse = await wcbtApi.fetch(
```
Shouldn't we first check if the latest version is already installed and skip the installation if so?

Stojdza replied:
It's not that I didn't think about it, but installing and activating WC via API was fast and I decided to go without checking if the specific WC version is already installed and activated. Additionally, we wouldn't run tests against external envs very often, so that was one more reason to skip the check. 

Anyway, I implemented it here [c9da3bd](https://github.com/woocommerce/woocommerce/pull/55063/commits/c9da3bddf46ed1cd95c55d86d328cec514df28fd). It took me longer than expected, but it should work now. When testing, please consider these scenarios and let me know your thoughts:

- Install the latest release with `pnpm test:e2e:pressable basic.spec.js` when:
   - the latest is already installed and activated
   - other version of WC is installed and activated
   - no WC is activated
- Install specific WC with `WC_VERSION=9.6.0-rc.1 pnpm test:e2e:pressable basic.spec.js` when:
   - the same version is already installed and activated
   - the latest is already installed and activated
   - no WC is activated

Thanks 🙌

Stojdza commented:
> I'm not seeing why this is needed. Both site-setup and reset should work fine. Can you point me to such a failure?

[Here](https://github.com/woocommerce/woocommerce/actions/runs/13203501453/job/36862124874#step:9:121) it is. `[reset]` was executed right after `[install wc]`. After I changed the order it passed, but I'm open to other suggestions.

adimoldovan commented:
> > I'm not seeing why this is needed. Both site-setup and reset should work fine. Can you point me to such a failure?
> 
> [Here](https://github.com/woocommerce/woocommerce/actions/runs/13203501453/job/36862124874#step:9:121) it is. `[reset]` was executed right after `[install wc]`. After I changed the order it passed, but I'm open to other suggestions.

Ah, I see. It was actually requiring global authentication, because it was missing the `.state/admin.json` file. BUT, it should not need that file, the `setup.use( { storageState: ADMIN_STATE_PATH } );` was probably a left-over and should be removed. Can you please remove it, and if you revert the dependencies changes I'm confident everything will work.

Commit:
Revert "Add dependencies for WPCOM reset"

This reverts commit 42cf1c71445ae5cca39117e15cce7772b6cac0d1.

Commit:
Revert "Add dependencies for Pressable reset"

This reverts commit 070cadd1254848009e0b8bb3960ac026e9b56c10.

Commit:
Remove ADMIN_STATE_PATH from reset.setup.js

Commit:
Fix spacing

Commit:
Revert "Revert "Add dependencies for Pressable reset""

This reverts commit 3b391757e8049709946281f5de6f94600ee3845a.

Commit:
Revert "Revert "Revert "Add dependencies for Pressable reset"""

This reverts commit ec99f26f1ef7dc9ec44e35eea066fd2208313139.

Commit:
Check if the given version is already installed and activated

Commit:
Check if the given version is already installed and activated

Commit:
Merge branch 'trunk' into e2e/preinstall-wc

Commit:
Wait for `site setup` before `reset`

Stojdza commented on plugins/woocommerce/tests/e2e-pw/envs/default-pressable/playwright.config.js:
```diff
@@ -16,6 +16,7 @@ config = {
 			name: 'reset',
 			testDir: `${ config.TESTS_ROOT_PATH }/fixtures`,
 			testMatch: 'reset.setup.js',
+			dependencies: [ 'site setup' ],
```
@adimoldovan, I added this dependency because reset was returning `500` without it ([example](https://github.com/woocommerce/woocommerce/actions/runs/13266737356/job/37036060223)). This way, it will wait for `setupProjects` to finish first. Please, let me know if there's a better idea on how to solve it. [Here](https://github.com/woocommerce/woocommerce/actions/runs/13287596105/job/37101904978) is the successful site reset when this change was added.

adimoldovan replied:
I could also reproduce this in my tests, and my best guess is that it happens because the plugin is not yet ready. Delaying the reset call helps and I'm not opposed to that, I'm only opposing it running after the site-setup, because it can potentially remove things that are set by the site-setup. In my mind it should be reset, then authentication, then site-setup.
In my testing I found that adding a simple check via the API that the correct plugin version is actually installed and active as mentioned [here](https://github.com/woocommerce/woocommerce/pull/55063#discussion_r1954300655) added enough delay for the reset to work. If not, I would be ok with adding 'global authentication' as a dependency for reset.

Stojdza replied:
After checking WC version at the end of the script [49422a7](https://github.com/woocommerce/woocommerce/pull/55063/commits/49422a742b5ef7b4c9cadcda3fefe7f5a3758203) I also reverted this change  [49422a7](https://github.com/woocommerce/woocommerce/pull/55063/commits/49422a742b5ef7b4c9cadcda3fefe7f5a3758203). Reset works fine now 👍

Commit:
Merge branch 'trunk' into e2e/preinstall-wc

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/fixtures/install-wc.setup.js:
```diff
@@ -0,0 +1,144 @@
+		setup.skip(
+			resolvedVersion === activatedWcVersion,
+			'Skip installing WC: The latest version is already installed and activated.'
+		);
```
I would no use `setup.skip` at this point, it already executed half of the setup and leads to confusion. The step was marked as skipped while the log was telling me `INSTALL_WC is enabled. Running installation script...`

```suggestion
		if(resolvedVersion === activatedWcVersion) {
			console.log('Required version is already installed and active.');
			return;
		}
```

Stojdza replied:
You're right. Logging was confusing indeed. I removed `setup.skip` from two places and improved logging here [c15600c](https://github.com/woocommerce/woocommerce/pull/55063/commits/c15600c9d45e2c1ff89521e4bc74fb5b4eda6103)

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/fixtures/install-wc.setup.js:
```diff
@@ -0,0 +1,144 @@
+		);
+	}
+
+	console.log( 'Installing with WC Beta Tester is finished.' );
```
Can we also add a check that the plugin version is actually installed and active, similar to the initial `wp-json/wp/v2/plugins` check.

Stojdza replied:
Added [49422a7](https://github.com/woocommerce/woocommerce/pull/55063/commits/49422a742b5ef7b4c9cadcda3fefe7f5a3758203)

adimoldovan reviewed (changes requested):
Thanks for the updates @Stojdza.

I left comments on code.

As a general comment, for a follow-up, not necessarily in this PR, we could address the fact that the deactivated versions are not deleted and different versions are piling up. We could [delete them](https://developer.wordpress.org/rest-api/reference/plugins/#delete-a-plugin).

Commit:
Introduce and use `getActivatedWooCommerceVersion` function

Commit:
Revert `dependencies: [ 'site setup' ]`

Commit:
Don't use `setup.skip` and improve logging to be more clear

Stojdza commented:
> As a general comment, for a follow-up, not necessarily in this PR, we could address the fact that the deactivated versions are not deleted and different versions are piling up. We could [delete them](https://developer.wordpress.org/rest-api/reference/plugins/#delete-a-plugin).

I agree. Let's address this in a separate PR 👍

Commit:
Merge branch 'trunk' into e2e/preinstall-wc

adimoldovan reviewed (approved):
Thanks for addressing all the suggestions @Stojdza! 🚀
